pkg/bucketaccess/bucketaccess_controller.go

```diff
@@ -0,0 +1,248 @@
+/* Copyright 2021 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucketaccess
+
+import (
+	"context"
+	"os"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	kubeerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+	kube "k8s.io/client-go/kubernetes"
+	kubecorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	buckets "sigs.k8s.io/container-object-storage-interface-api/clientset"
+	bucketapi "sigs.k8s.io/container-object-storage-interface-api/clientset/typed/objectstorage.k8s.io/v1alpha1"
+	cosi "sigs.k8s.io/container-object-storage-interface-spec"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// BucketAccessListener manages Bucket objects
+type BucketAccessListener struct {
+	provisionerClient cosi.ProvisionerClient
+	provisionerName   string
+
+	kubeClient   kube.Interface
+	bucketClient buckets.Interface
+	kubeVersion  *utilversion.Version
+
+	namespace string
+}
+
+// NewBucketAccessListener returns a resource handler for BucketAccess objects
+func NewBucketAccessListener(provisionerName string, client cosi.ProvisionerClient) (*BucketAccessListener, error) {
+	ns := os.Getenv("POD_NAMESPACE")
+	if ns == "" {
+		return nil, errors.New("POD_NAMESPACE env var cannot be empty")
+	}
+
+	return &BucketAccessListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+		namespace:         ns,
+	}, nil
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
```

Comment:
Fix comment: this func grants access to an existing bucket

Reply:
fixed